src/libraries/System.Private.CoreLib/src/System/String.cs

```diff
@@ -509,6 +509,25 @@ internal static string CreateFromChar(char c1, char c2)
             return result;
         }
 
+        internal static string CreateFromSpan(ref char buffer, int length)
+        {
+            Debug.Assert(length >= 0);
+
+            if (length == 0)
+            {
+                return Empty;
+            }
+
+            string result = FastAllocateString(length);
+
+            Buffer.Memmove(
+                elementCount: (uint)result.Length, // derefing Length now allows JIT to prove 'result' not null below
```

Comment:
I'm not understanding the comment. If the JIT can prove that result is not-null here, I'd expect both this and the below call to not have null checks. If it can't, I'd expect a null check here and not below, and if you use `length` instead, I'd expect just a null check below... either way, one check. How does reality differ from my expectations?

Reply:
Normally the parameters are passed `(dest, src, len)`, and they're evaluated left-to-right. So without any named arguments, this would naturally read:

```cs
Buffer.Memmove(ref result._firstChar, ref buffer, (uint)result.Length);
```

The problem with this is that the JIT can't prove that _result_ (the return value from `FastAllocateString` is non-null. So the expression `ref result._firstChar`, instead of being just one simple address calculation instruction, actually performs two pieces of work: (a) first check if _result_ is null and trigger an AV (NRE) if it is; then (b) perform the address calculation.

By moving the `result.Length` expression first, the JIT knows that _result_ cannot be null, otherwise the dereference would have failed and control flow wouldn't have gotten this far. So it can skip the pessimistic check described in (a) above and perform the address calculation without any further ceremony.

To be fair, the null check described in (a) isn't that expensive. But it's unnecessary and relatively easy to avoid by swapping parameter evaluation order, so I saw it as low-hanging fruit.

Reply:
I'm not objecting, but trying to understand. Are you saying your version is more efficient than:
```
Buffer.Memmove(ref result._firstChar, ref buffer, (uint)length);
```
Why does the JIT need more null checks for that?

Reply:
If this was written in natural way, the codegen for this would be:
```
test [result], result // null check result 
lea arg0, [result + 0xC]
mov arg1, ...
mov arg2, [result + 0x8] // length
```
With this tweak, it will look like this:
```
mov arg2, [result + 0x8] // length
lea arg0, [result + 0xC] // null check on result is not needed because of it was done as side-effect of loading length
mov arg1, ...
```
So this is saving one instruction. Non-measurable improvement. One can argue whether tweaks like this add valuable - they make the code cryptic.

There is nothing preventing the JIT to do this optimizations for you. The current JIT is just not smart enough to do it.

Reply:
Btw, maybe it makes sense to mark some BCL methods with "I swear I can't return null" type of attribute to give jit hints? E.g. we give such hints to our LLVM back-ends for some things (e.g. managed allocators). `FastAllocateString` - nonnull, noalias.

Reply:
@EgorBo Do you know of any outstanding issues tracking this? I'd like to be able to provide JIT hints: "this can't be null" or "this is always non-negative" or whatever.

Reply:
@GrabYourPitchforks none I am aware of

Reply:
https://github.com/dotnet/runtime/issues/4966